modules/pallets/token-gateway/src/lib.rs

```diff
@@ -618,58 +618,73 @@ where
+				let multi_signature = MultiSignature::decode(&mut &*signature)
+					.map_err(|err| anyhow!("Signature decode error: {err:?}"))?;
+
+				// Verify signature against encoded runtime call
+				let nonce = frame_system::Pallet::<T>::account_nonce(beneficiary.clone());
+				let payload = (nonce, substrate_data.runtime_call.clone()).encode();
+				let message = sp_io::hashing::keccak_256(&payload);
+
+				match multi_signature {
+					MultiSignature::Ed25519(sig) => {
+						let pub_key = body.to.0.as_slice().try_into().map_err(|_| {
+							anyhow!("Failed to decode beneficiary as Ed25519 public key")
+						})?;
+						if !sp_io::crypto::ed25519_verify(&sig, message.as_ref(), &pub_key) {
+							Err(anyhow!(
 							"Failed to verify ed25519 signature before dispatching token gateway call"
 						))?
-					}
-				},
-				MultiSignature::Sr25519(sig) => {
-					let pub_key = body.to.0.as_slice().try_into().map_err(|_| {
-						anyhow!("Failed to decode beneficiary as Sr25519 public key")
-					})?;
-					if !sp_io::crypto::sr25519_verify(&sig, message.as_ref(), &pub_key) {
-						Err(anyhow!(
-							"Failed to verify sr25519 signature before dispatching token gateway call"
-						))?
-					}
-				},
-				MultiSignature::Ecdsa(sig) => {
-					let pub_key = sp_io::crypto::secp256k1_ecdsa_recover(&sig.0, &message)
-						.map_err(|_| {
-							anyhow!("Failed to recover ecdsa public key from signature")
+						}
+					},
+					MultiSignature::Sr25519(sig) => {
+						let pub_key = body.to.0.as_slice().try_into().map_err(|_| {
+							anyhow!("Failed to decode beneficiary as Sr25519 public key")
 						})?;
-					let eth_address =
-						H160::from_slice(&sp_io::hashing::keccak_256(&pub_key[..])[12..]);
-					let substrate_account = T::EvmToSubstrate::convert(eth_address);
-					if substrate_account != beneficiary {
-						Err(anyhow!(
-							"Failed to verify signature before dispatching token gateway call"
+						if !sp_io::crypto::sr25519_verify(&sig, message.as_ref(), &pub_key) {
+							Err(anyhow!(
+							"Failed to verify sr25519 signature before dispatching token gateway call"
 						))?
-					}
-				},
-			}
+						}
+					},
+					MultiSignature::Ecdsa(sig) => {
+						let pub_key = sp_io::crypto::secp256k1_ecdsa_recover(&sig.0, &message)
+							.map_err(|_| {
+								anyhow!("Failed to recover ecdsa public key from signature")
+							})?;
+						let eth_address =
+							H160::from_slice(&sp_io::hashing::keccak_256(&pub_key[..])[12..]);
+						let substrate_account = T::EvmToSubstrate::convert(eth_address);
+						if substrate_account != beneficiary {
+							Err(anyhow!(
+								"Failed to verify signature before dispatching token gateway call"
+							))?
+						}
+					},
+				};
+
+				beneficiary.clone().into()
+			} else {
+				// check if the first 12 bytes of body.from is all zeroes
+				if body.from[0..12] == [0u8; 12] {
+					// sender is evm account
+					T::EvmToSubstrate::convert(H160::from_slice(&body.from[12..]))
```

Comment:
Is it worth it to enforce that the source chain must be an evm chain before doing this?

Reply:
I also want to substrate chains